### PR TITLE
Terraform Cloud Kubernetes Operator v2 Migration Guide

### PR DESCRIPTION
**Pull request by BrianMMcClain**
### What

:mag: [Deploy Preview](https://terraform-docs-common-git-bmm-k8s-op-migration-hashicorp.vercel.app/terraform/cloud-docs/integrations/kubernetes/k8s-ops-v2-migration)
Migration guide from v1 to v2 of the Terraform Cloud Kubernetes Operator

### Why

With v2 of the operator in beta, we want to provide a way for v1 users to migrate to v2

----------

### Merge Checklist
_If items do not apply to your changes, add (N/A) and mark them as complete._

#### Pull Request
- [x] One or more labels describe the type of change (e.g. clarification) and associated product (e.g. tfc).
- [x] Description links to related pull requests or issues, if any.

#### Content
- [x] Redirects have been added to `website/redirects.js` for moved, renamed, or deleted pages.
- [x] API documentation and the API Changelog have been updated. 
- [x] Links to related content where appropriate (e.g., API endpoints, permissions, etc.).
- [x] Pages with related content are updated and link to this content when appropriate.
- [x] Sidebar navigation files have been updated for added, deleted, reordered, or renamed pages.
- [x] New pages have metadata (page name and description) at the top.
- [x] New images are 2048 px wide. They have HashiCorp standard annotation color (#F92672) and format (rectangle with rounded corners), blurred sensitive details (e.g. credentials, usernames, user icons), and descriptive alt text in the markdown for accessibility.
- [x] New code blocks have the correct syntax and line breaks to eliminate horizontal scroll bars.
- [x] UI elements (button names, page names, etc.) are bolded.
- [x] The Vercel website preview successfully deployed.

#### Reviews
- [x] I or someone else reviewed the content for technical accuracy.
- [x] I or someone else reviewed the content for typos, punctuation, spelling, and grammar.

### CODE REVIEW

File: website/docs/cloud-docs/integrations/kubernetes/k8s-ops-v2-migration.mdx [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+Next, backup the workspace resources.
+
+```shell-session
+$ kubectl get workspace --all-namespaces -o yaml > backup_tfc_operator_v1.yaml
+```
+
+Finally, we suggest that you set the `applyMethod` of each workspace to `manual`. In version 2 of the Terraform Cloud Kubernetes Operator this is the default value, however explicitly configuring it with the value provides extra protection from unintended automatic applies during the upgrade process.
+
+```shell-session
+kubectl patch workspace <WORKSPACE_NAME> --type=merge --patch '{"spec": {"applyMethod": "manual"}}'
```
Comment (robin-norwood) [inline website/docs/cloud-docs/integrations/kubernetes/k8s-ops-v2-migration.mdx:29]: So they run this for each of their TFC workspaces? Or the ones that use kubernetes? Can we give them more specific guidance here?

Comment (robin-norwood) [reply website/docs/cloud-docs/integrations/kubernetes/k8s-ops-v2-migration.mdx:29]: Or does this refer to kubernetes workspaces?

File: website/docs/cloud-docs/integrations/kubernetes/k8s-ops-v2-migration.mdx [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+```
+## Upgrade the operator
+
+Clone the Terraform Cloud Kubernetes Operator repository.
```
Comment (robin-norwood) [inline website/docs/cloud-docs/integrations/kubernetes/k8s-ops-v2-migration.mdx:33]:
```suggestion
Clone the Terraform [Cloud Kubernetes Operator repository](https://github.com/hashicorp/terraform-cloud-operator).
```

File: website/docs/cloud-docs/integrations/kubernetes/k8s-ops-v2-migration.mdx [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+$ kubectl diff -f workspaces_patch_a.yaml
+```
+
+Patch the workspace CRD with **patch A**. This patch adds `app.terraform.io/v1alpha2` support, but excludes `.status.runStatus` because it has a different format in `app.terraform.io/v1alpha1` and causes JSON unmarshalling issues.
+
+```shell-session
+$ kubectl patch crd workspaces.app.terraform.io --patch-file workspaces_patch_a.yaml
+```
+
+Follow the [installation guide](https://github.com/hashicorp/terraform-cloud-operator/blob/main/docs/installation.md) to install v2 of the operator.
```
Comment (robin-norwood) [inline website/docs/cloud-docs/integrations/kubernetes/k8s-ops-v2-migration.mdx:57]: So we're supposed to leave this guide and follow the steps in that one? That page has a few different methods to install - should I follow the `install` command or the `upgrade` command? I think this needs a little more explicit guidance.

File: website/docs/cloud-docs/integrations/kubernetes/k8s-ops-v2-migration.mdx [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+Next, create a Kubernetes secret to store the TFC API token following the [Usage Guide](https://github.com/hashicorp/terraform-cloud-operator/blob/main/docs/usage.md#prerequisites). The API token can be copied from the Kubernetes secret that you created for v1 of the operator. By default, this is named `terraformrc`. Use the `kubectl get secret` command to get the API token.
+
+```shell-session
+$ kubectl --namespace ${RELEASE_NAMESPACE} get secret terraformrc -o json | jq '.data.credentials' | tr -d '"' | base64 -d
```
Comment (robin-norwood) [inline website/docs/cloud-docs/integrations/kubernetes/k8s-ops-v2-migration.mdx:62]: Some question about setting the `RELEASE_NAMESPACE` variable

File: website/docs/cloud-docs/integrations/kubernetes/k8s-ops-v2-migration.mdx [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+```
+
+If v2 of the operator pod fails to start, ensure that each `workspace` object has the `spec.token` value set. To do this, patch each `workspace` object using the `kubectl patch` command.
+
+```shell-session
+$ kubectl patch workspace <WORKSPACE_NAME> --type=merge --patch '{"spec": {"terraformVersion": "1.4.6", "token": {"secretKeyRef": {"key": "token", "name": "tfc-operator"}}}}'
```
Comment (robin-norwood) [inline website/docs/cloud-docs/integrations/kubernetes/k8s-ops-v2-migration.mdx:68]: This is again iterating over each of their workspaces, right?

File: website/docs/cloud-docs/integrations/kubernetes/k8s-ops-v2-migration.mdx [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+```
+
+Validate
```
Comment (robin-norwood) [inline website/docs/cloud-docs/integrations/kubernetes/k8s-ops-v2-migration.mdx:71]: "Validate that..."

File: website/docs/cloud-docs/integrations/kubernetes/k8s-ops-v2-migration.mdx [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+
+```shell-session
+$ kubectl logs -f <POD_NAME>
+```
```
Comment (robin-norwood) [inline website/docs/cloud-docs/integrations/kubernetes/k8s-ops-v2-migration.mdx:75]: What output do we expect to see?

File: website/docs/cloud-docs/integrations/kubernetes/k8s-ops-v2-migration.mdx [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+Specifically, look for an error message such as the following.
+
+```
+ERROR	Migration	{"workspace": "default/<WORKSPACE_NAME>", "msg": "spec contains old finalizer finalizer.workspace.app.terraform.io"}
+```
+The `finalizer` exists to provide greater control over the migration process. Verify the custom resource, and when you’re ready to migrate it, use the `kubectl patch` command to update the `finalizer` value.
+
+```shell-session
+$ kubectl patch workspace migration --type=merge --patch '{"metadata": {"finalizers": ["workspace.app.terraform.io/finalizer"]}}'
+```
```
Comment (robin-norwood) [inline website/docs/cloud-docs/integrations/kubernetes/k8s-ops-v2-migration.mdx:104]: Are we done? Do we need to re-run the `kubectl patch` command if it failed the first time bc of the v1 finalizer?

Regardless, I think we should say something to let the user know they are done, and maybe validate that it worked?

File: website/docs/cloud-docs/integrations/kubernetes/k8s-ops-v2-migration.mdx (hunk 6 above)
Comment (BrianMMcClain) [reply website/docs/cloud-docs/integrations/kubernetes/k8s-ops-v2-migration.mdx:71]: Oops, good catch 😆

File: website/docs/cloud-docs/integrations/kubernetes/k8s-ops-v2-migration.mdx (hunk 3 above)
Comment (BrianMMcClain) [reply website/docs/cloud-docs/integrations/kubernetes/k8s-ops-v2-migration.mdx]: Pulled the direct install command from the guide, need to verify with eng that this is the intended use.

File: website/docs/cloud-docs/integrations/kubernetes/k8s-ops-v2-migration.mdx (hunk 7 above)
Comment (BrianMMcClain) [reply website/docs/cloud-docs/integrations/kubernetes/k8s-ops-v2-migration.mdx:106]: /cc @arybolovlev

File: website/docs/cloud-docs/integrations/kubernetes/k8s-ops-v2-migration.mdx [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+$ kubectl diff -f workspaces_patch_a.yaml
+```
+
+Patch the workspace CRD with **patch A**. This patch adds `app.terraform.io/v1alpha2` support, but excludes `.status.runStatus` because it has a different format in `app.terraform.io/v1alpha1` and causes JSON unmarshalling issues.
+
+```shell-session
+$ kubectl patch crd workspaces.app.terraform.io --patch-file workspaces_patch_a.yaml
+```
+
+Install v2 the Helm chart with the `helm install` command. Be sure to set the `operator.watchedNamespaces` value to the list of namespaces your Workspace resources are deployed to.
```
Comment (arybolovlev) [inline website/docs/cloud-docs/integrations/kubernetes/k8s-ops-v2-migration.mdx:76]: I like that we have mentioned the ability to watch specific namespaces, this is one of the options that practitioners were waiting for. Will it be worth mentioning that without this option the operator will watch all namespaces?

File: website/docs/cloud-docs/integrations/kubernetes/k8s-ops-v2-migration.mdx (hunk 7 above)
Comment (arybolovlev) [reply website/docs/cloud-docs/integrations/kubernetes/k8s-ops-v2-migration.mdx:106]: The operator produces quite a lot of logs, and they are all either `INFO` or `ERROR`.

The reconciliation loop always starts with the following message:

```
INFO	Workspace Controller	{"workspace": {"name":"<WORKSPACE_NAME>","namespace":"<NAMESPACE>"}, "msg": "new reconciliation event"}
```

When everything goes well, we should see only `INFO` messages with the final message:

```
INFO	Workspace Controller	{"workspace": {"name":"<WORKSPACE_NAME>","namespace":"<NAMESPACE>"}, "msg": "successfully reconcilied workspace"}
```

In the case the operator encounters any issues, we would see an `ERROR` message, it has the following format:

```
ERROR	Workspace Controller	{"workspace": {"name":"<WORKSPACE_NAME>","namespace":"<NAMESPACE>"}, "msg": "<ERROR_MESSAGE>"}
```

Usually, the error message is quite informative and it shouldn't be a problem to understand it and fix it. Live example:

```
ERROR	Workspace Controller	{"workspace": {"name":"this","namespace":"default"}, "msg": "failed to get terraform cloud client", "error": "Secret \"tfc-operator\" not found"}
```

In short, we expect to see only `INFO` messages.

**Review by arybolovlev — APPROVED** [review]
Looks great, @BrianMMcClain! I have added one proposal regarding the watched namespaces. 🚀

Thank you! 👍🏻

File: website/docs/cloud-docs/integrations/kubernetes/k8s-ops-v2-migration.mdx (hunk 9 above)
Comment (BrianMMcClain) [reply website/docs/cloud-docs/integrations/kubernetes/k8s-ops-v2-migration.mdx:76]: Added, thanks!

**Review by robin-norwood — APPROVED** [review]
lgtm!
